### PR TITLE
[FIX] product, website_sale: improve the working of the remove button

### PR DESCRIPTION
**Pull request by loco-odoo**
[FIX] web_editor: do not create variant image while selecting an image
Steps to reproduce the bug:
- Go on a product page of a product with multiple variants but no image
set for those variants. Because there is no image set to those variants,
the variant images fall back on the template image.
- Edit.
- Click on the variant image of the product. (Note that `.o_dirty` is
added to the element).
- Save.

-> The image of the product is not the template image anymore but a
variant image that is the same as the template image.

Because `.o_dirty` is added while clicking on the product image,
`save_embedded_field()` will set the `image_1920` field of
`product.product`. Because the product has multiple variants, the
`image_variant_1920` field of `product.product` is modified see
(`_set_template_field()`).

The problem is that the `o_dirty` class is added on the clicked element.
To solve the problem, the `MutationObserver` responsible for adding this
class is paused when modifying the tooltips. Note that it was introduced
by [1] but it was broken since [2].

[1]: https://github.com/odoo/odoo/commit/9f93fa8e77b11da2fcf60f606784eec49a94778a
[2]: https://github.com/odoo/odoo/commit/8eb0ca54200f0f8e1078bb6a0d507e747ccea122

task-3111601

-----------------------------------------------------------------------------------------------------------------------------------------------------------------

[FIX] website_sale: improve the working of the remove button

Steps to reproduce the bug:
- Open a product page in edit mode.
- Replace the main image with one of your own.
- Click on the "Remove" button.

-> Nothing happens and the "Remove" button has to be clicked one more
time in order to work properly.

The goal of this commit is to remove the need of the double clicking
operation in order to remove a media. Before this commit, here was the
process that happened at the removal of the main product image:
- `'/shop/product/remove-image'` was called and the field `image_1920`
of the record (`product.product`) was set to `False`.
- A `request_save` was triggered up.
- At the "request save", the `Wysiwyg` performs two main actions:
[1] `saveModifiedImages()`: Using the `o_modified_image_to_save` class,
the `Wysiwyg` saves the updated version of the image and sets its new
attachment URL in the `src` attribute before completing the save
process.
[2] `_saveViewBlocks()`: in `save_embedded_field()`, the field
`image_1920` of `product.product` is updated tanks to the value saved in
the new created attachment.

The problem here is double: first, a useless attachment is created while
the user wants to delete the image. Second, at the end of the process,
the field `image_1920` of the `product.product` is not set to `False`
anymore.
To resolve this problem, the image is removed from the DOM before the
"request save". Thanks to that, no attachment is created at the "request
save". Moreover, because the image is removed, the field `image_1920` of
`product.product` is set to `False` at `save_embedded_field()` (see
`Image.from_html()`). Note that the rpc call to `remove_product_image()`
has been removed. Indeed, the modification of the `image_1920` field of
the records of type `product.product` and `product.template` is now
handled in `save_embedded_field()`. The `unlink()` of the
`product.image` is kept because the record has to be deleted at the
remove of a secondary image.

task-3111601

### CODE REVIEW

**Comment by loco-odoo** [comment]
Reviewed by Guillaume (gdi)

**Comment by qsm-odoo** [comment]
@robodoo delegate=xO-Tx

**Review by xO-Tx — CHANGES_REQUESTED** [review]
@loco-odoo It seems that removing the `'.o_modified_image_to_save'` doesn't handle all cases... the use of an external URL creates the attachment too... e.g.

https://user-images.githubusercontent.com/25032876/221619912-24b78bb8-8a55-44f9-82b3-961674c06c1a.mp4


Can you check this :+1:

**Comment by loco-odoo** [comment]
@xO-Tx, Yes, it definitely seems to have a bigger issue here as you mention it! I agree with all of what you said in your explanation, however, may I suggest an other idea in order to make it work?
IMO, the problem is that the field value of `"product.product"` is modified two times when removing an image : 
- Once at the call of `"/shop/product/remove-image"` (the field `image_1920` of `"product.product"` is set the `False` and this is what we want).
- An other time, in `"save_embedded_fields"` at the `"request_save"`.
IMO, the second field modification is useless and could be avoided by removing the `o_dirty` class. In this case, we would have something that looks like this : 
```diff
     async removeMedia() {
         this._rpc({
             route: '/shop/product/remove-image',
             params: {
                 image_res_model: this.recordModel,
                 image_res_id: this.recordId,
             },
-        }).then(() => this.trigger_up('request_save', {reload: true, optionSelector: '#product_detail_main'}));
+        }).then(() => {
+            this.$target[0].classList.remove("o_modified_image_to_save");
+            this.$target[0].parentElement.classList.remove("o_dirty");
+            this.trigger_up('request_save', {reload: true, optionSelector: '#product_detail_main'});
+        });
     },
```
The idea is : 
- To remove the `o_modified_image_to_save` class because we do not want to create an attachment for an image when this image is going to be deleted.
- To remove the `o_dirty` class to not update the field of `"product.product"` at the `"request_save"`. Indeed, it has already been modified at the call of `"/shop/product/remove-image"`.
What do you think ?

**Comment by qsm-odoo** [comment]
> I did this move because I asked specifically to Pierre if it would be possible to create this task as I reported him a bug but found a possible fix. So I directly assigned it to me to prevent someone else from spending time on it.

(task comment) @loco-odoo Next time, let's let the website_sale team handle their own bug fixes, we have enough with our own at the moment :smile: 

--

@loco-odoo @xO-Tx Thanks for your investigations, to be sure I understand everything:
- The "remove" button on shop images does not work when it is about the main product image
- This is because what the code do is "remove image, then save page... which saves the same image again"
- Mehdi suggests "remove -> save (add again) -> remove again", which is not bad as we would not bypass the normal save which could do some more important things... but that's seems hackish/inefficient/etc
- Louis suggests "remove -> then does not save the area (not dirty anymore)" -> better to me, but the concerns of Mehdi are still right.

From the main problem "The "remove" button on shop images does not work when it is about the main product image", I actually immediately thought "Why are we even allowing to remove the product main image?". IMO if the problem is really that, I would just hide the remove button when it concerns the main image, I don't see the point in removing the main image. That would need PO validation though (there is a new PO, don't remember its quadrigram). If not (we either don't want that or I misunderstood and the remove button problem does not concern the main image only), I would rather go with:
- Remove this `remove-image` rpc
- Clicking the remove button actually removes the image in the DOM
- request_save
- adapt `save_embedded_field` to rightfully set the image field to `False` since it was removed

**Comment by loco-odoo** [comment]
> the remove button problem does not concern the main image only

Unfortunately yes, the problem appears also for extra images (see video)

[Bug_remove_secondary_images.webm](https://user-images.githubusercontent.com/114914971/222777265-f116c2d1-1c3f-48d3-a5ce-d31bd7b88757.webm)

**Comment by loco-odoo** [comment]
> The "remove" button on shop images does not work when it is about the main product image

Unfortunately, It also does not work for extra images (see video). However, in this case, it seems that the error is not completely the same than for the primary image. In the case of an extra image (the model is `product.image`), the call to `"/shop/product/remove-image"` will lead to the call to `unlink()` on `product.image`.  As a result, the attachments bound to the `product.image` will be deleted. However, the image that we want to delete has the `o_modified_image_to_save` class so there will be a call to `saveModifiedImages()` (to try to create an attachment) at the `"request_save"`. The problem is that to create a new attachment, a copy of the original one is made and then modified according the data of the new image. Because the original attachment as been deleted, an error is thrown.

> adapt `save_embedded_field` to rightfully set the image field to `False` since it was removed

@qsm-odoo , so for the fix in stable, your suggestion is to modify the function of a python file?


[Bug_extra_img.webm](https://user-images.githubusercontent.com/114914971/223089113-5d912ebf-1313-4a78-a944-1ca0baa37d05.webm)

**Comment by qsm-odoo** [comment]
> Unfortunately, It also does not work for extra images (see video).

Strange I did not had the issue when trying it out. Maybe because I was doing it during the same edition? In any case strange that those buttons were merged when did not work in most cases... :grimacing: 

> so for the fix in stable, your suggestion is to modify the function of a python file?

Well also the JS (step 2 "Clicking the remove button actually removes the image in the DOM") but yes, would that be a problem? I think it would solve both the main product image case and the product.image case you described, no? Would also be better in terms of usability: the image would only be removed if the save actually succeeds. Technically, the solution would also be generic and allow to empty any image field in the future (not only product.image).

**Comment by loco-odoo** [comment]
> but yes, would that be a problem?

Actually not, my mistake came from the fact that I thought that we could not modify python file in stable. However, I think that since the changes will not affect the database schema, it is ok to make those changes in stable (but correct me if I am wrong :relaxed: )

**Comment by qsm-odoo** [comment]
@loco-odoo Yes, changing python code which does not the affect the database schema is the most stable thing you can do in general: you change something -> the user gets the update next time it simply *restarts* its server (once a week at least on the saas).

File: addons/website_sale/controllers/main.py
Comment (xO-Tx) [inline addons/website_sale/controllers/main.py:567]: Which case is handled by this code if we always submit a `'product.product'` or `'product.image'` field?

**Review by xO-Tx — CHANGES_REQUESTED** [review]
I'm not sure we can rely 100% on field update to remove the target in all cases... but if you:

- Go to a product page (with multiple variants)
- Select the main image > Add extra images...
- select the 'product.product' image again.
- Remove > The image is still there...


https://user-images.githubusercontent.com/25032876/223716140-1d5cd222-a76d-4eb5-b96c-bb8d11cc74ea.mp4



I think this had something to do with the way the 'product.product' image field is computed No?

**Comment by loco-odoo** [comment]
@xO-Tx , indeed, the example that you showed in your example was not handled by the code. In your example, the product has multiple variants so the image will be stored at the `image_variant_1920` field of the `product.product` record. In order to work properly, this field is now set to `False` in the `remove_product_image` function call. With this in hand, the `product_product._get_images()` function had to be adapted in order to avoid the case where the empty main image is placed at the end of all the other images in the carousel thumbnail (resulting in an undeletable last empty image).

File: addons/website_sale/controllers/main.py
Comment (loco-odoo) [reply addons/website_sale/controllers/main.py:567]: Indeed, I do not see neither which case is handled here as, if I am not mistaken, it is impossible to directly remove a `product.template` (It is possible to modify it but not delete it (see picture))
image

**Review by xO-Tx — APPROVED** [review]
LGTM thanks @loco-odoo for the updates!
@qsm-odoo some remarks before merge:

1- Apparently dead code here: https://github.com/odoo/odoo/pull/108684/commits/792f6ab5c7d722190aa1d7f1fc4c24a0706abdfc#diff-83686977d1607bdb0a3e755e9d22ec4496c1dc4caebfff4bee3902ddd1b3e866L566-L567 still don't know why it was added in first place (Even after I checked the original commit where it was added: https://github.com/odoo/odoo/commit/3958d15086bc7b3f5af7b3da83562b152b6dfc7c) 
2-  The order of images updated here: https://github.com/odoo/odoo/pull/108684/commits/792f6ab5c7d722190aa1d7f1fc4c24a0706abdfc#diff-bcce8086c8655fe6671ccbb5b4739a891993e92380f67930ea24423e80941607R78-R80 to set the empty `product.product` image as first element in carousel when the `product.template` image is also empty (users with both images empty will see the empty varian image first instead of product extra images...)

So do you think we need an eComerce approval for 1 & 2? or it's OK :+1:

File: addons/website_sale/controllers/main.py
Comment (qsm-odoo) [inline addons/website_sale/controllers/main.py:570]: What about not calling this controller at all when trying to remove the main image (that was I was trying to say in my previous message) ? :wink:

File: addons/website_sale/models/product_product.py
Comment (qsm-odoo) [inline addons/website_sale/models/product_product.py:80]: Not sure why this change? :thinking:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Commit message says
> the modification of the `image_1920` field of the records of type
`product.product` and `product.template` has been deleted. Indeed, this
is now handled in `save_embedded_field()`

But this is not what I am seeing: no patch of save_embedded_field and the route still modfies the product.product... :grimacing:

File: addons/website_sale/controllers/main.py
Comment (qsm-odoo) [reply addons/website_sale/controllers/main.py:570]: I'll need some more convincing, not sure to understand :grimacing: All I am asking is to delay this emptying of the image_variant_1920 field at the save lavel rather than with `remove-image` as a "pre-saving"... surely that should be possible :grimacing:

File: addons/website_sale/models/product_product.py
Comment (loco-odoo) [reply addons/website_sale/models/product_product.py:80]: This is to avoid the case of the picture. In this case, the products has multiple variants and its main image has been deleted. However, without this modification, the main image (that has been deleted) is placed at the end of all the secondary images. IMO, this is problematic for two reasons: First, this behavior is specific for products with variants. Second, the result is a last undeletable empty image (as it is the main image, it should not be possible to delete it)
image

File: addons/website_sale/controllers/main.py
Comment (qsm-odoo) [reply addons/website_sale/controllers/main.py:570]: @loco-odoo That's exactly my point: what is specific behavior should stay as specific code and what is generic should be implemented generically:
- Removing a product.image record -> that's a very specific feature (there is whole model dedicated to extra product images and records of those models have to be deleted/added to remove/add image via the editor) => this should stay as a RPC done on click of the button, then we have to save the whole page so that the page reloads with the proper rendering.
- Emptying the main product image -> if we want that feature (**), this should be a generic behavior: we have an image t-field that we want to empty. The editor could be in charge of "emptying that field" = removing the image / displaying the 404 image and the saving (`save_embedded_field`) could be in charge of actually save that "emptying of a field".

And even if we don't agree on that part: it is technically more consistent and "more efficient" if it is done that way:
- You save the page, the saving of the edited image is done at the same time, same as all other potentially edited t-field and other content in the page. + If the saving fails, then the saving fails: you don't have a half-save that has been done before the main save crashed.
- You avoid a RPC to be done in the case of removing the main product image (since you just let the normal RPC save to be done for that). That argument alone should be enough :wink: 

> I would just hide the remove button when it concerns the main image, I don't see the point in removing the main image. That would need PO validation though (there is a new PO, don't remember its quadrigram).

** Also, this :point_up: is still a possibility to me.

**Comment by AntoineVDV** [comment]
Hello guys, as I'm a total newbie with the editor I wanted to reproduce the issue reported here but everything seems to work fine on my side in 16.0 :thinking: 

Do you mind sharing a video of the bug? I must have missed something...

**Comment by loco-odoo** [comment]
> Do you mind sharing a video of the bug?

@AntoineVDV , sure, here it is :+1: 


[Bug_remove_main_product_image.webm](https://user-images.githubusercontent.com/114914971/227472027-937f1df8-1cc0-42c9-bb80-556c6fb8dfe2.webm)

File: addons/website_sale/controllers/main.py
Comment (qsm-odoo) [reply addons/website_sale/controllers/main.py:570]: > But isn't it consider as mixing specific code with generic code (as we would refer to product.product at the web_editor level)?

Exactly, that's why I did not imagine doing that but instead doing it generically :wink: In `save_embedded_field`, I imagined something like "if value/image is null, then the given field of the given model (you don't have to know it is product.product and image_variant_1920) is set to False".

Then a write of image_1920 to False should do the proper thing, whatever they are (if it is a computed field like you said (`_compute_image_1920`), then there could be an inverse function which does the proper things when written?).

> However, may I suggest also an other way of doing? [...] Sure, we can also see with the PO of ecommerce what he/she thinks about removing the possibility to remove a main image +1

Yes, at this point I think we really should have @AntoineVDV and its PO's opinion... :wink:

File: addons/product/models/product_product.py
Comment (loco-odoo) [inline addons/product/models/product_product.py:104]: Due to the update of the condition, `record[template_field]` could be removed here. But maybe is it better to let it (easier for the comprehension)?

File: addons/website_sale/tests/test_website_sale_image.py
Comment (qsm-odoo) [inline addons/website_sale/tests/test_website_sale_image.py:385]: With `setUp`, you can save those things in `self` for use in the tests :+1:

File: addons/website_sale/static/src/js/website_sale.editor.js
Comment (loco-odoo) [reply addons/website_sale/static/src/js/website_sale.editor.js]: Indeed, `Image.from_html()` already returns `False` if the image is not in the DOM :+1:

**Comment by loco-odoo** [comment]
@qsm-odoo , just to be sure that you still have this one in mind :+1:

File: addons/product/models/product_product.py
Comment (qsm-odoo) [inline addons/product/models/product_product.py:101]: @loco-odoo I still don't understand this change, I just tested, this just does not make sense :grimacing: Have a product with 10 variants, configure an image for the template (base case), then set up the image for like 3 variants. Then, enter edit mode remove the image you set up for one of those 3 variants => bam now 8 variants do not have an image anymore (the current one and the 7 others that had no specific image)... I really don't think this is what we want, we don't want the template image to be removed when editing a variant, ever, or I am missing something... :shrug:

There probably is still a spec problem about all this: if we keep the current 16.0 behavior, removing a variant image would revert it back to the template image (seems good) but... what should hitting the "remove" button again do? I don't know. Can you check with PAJU what we want for all use cases? :grimacing: 

- Do we show the "remove" button always? (last time she said yes but I don't think all usecases were considered (like the one I described above)).
- What does it to in all the possible use case:
    - Remove a variant's specific image
    - Remove a variant's image which is in fact the template image (while it is used by other variants or not)
    - Remove the template image (while it is used by many variants or not)
    - More ?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/product/models/product_product.py
Comment (loco-odoo) [reply addons/product/models/product_product.py]: @paju-odoo, can you confirm that this is what you want for the remove of the main image :

-For a product with no variants, the image of the product falls back on the image of the template so removing the image of the product leads to removing the image of the template (see video 1).

-For a product with multiples variants :

1. If there is no specific image for the variant : the product image of the variant falls back on the template image. Removing a product image variant leads to removing the template image → the template and all the other variants with no specific image loose their image (see video 2).
2. If there is a specific image for the variant, removing the product variant image leads to fallback on the template image. (see video 3). Note than clicking again on the "Remove button" leads to the same behavior than the one just above.

Thanks in advance!


[Screencast from 06-20-2023 05:25:36 PM.webm](https://github.com/odoo/odoo/assets/114914971/210b85b3-753b-44b1-9930-1745404f6d64)


[Screencast from 06-20-2023 05:27:33 PM.webm](https://github.com/odoo/odoo/assets/114914971/8024bd89-8d10-40f7-8d60-0dfd95caba9d)


[Screencast from 06-20-2023 05:31:04 PM.webm](https://github.com/odoo/odoo/assets/114914971/d7ac66b2-42b5-44c8-b621-a58da5915d44)

Comment (paju-odoo) [reply addons/product/models/product_product.py]: > 
@loco-odoo 

Indeed, that's the expected behavior and how it's been working so far. If no specific image has been set on a variant, the template's image is in fact the variant's image, so removing it will remove the image from the template and all other variants using it. 

I'm not saying this is the best behavior, but it is the current one and we should not try to change it in regards with this fix. 

We'll have a look into it in the eCommerce and try to come up with a better solution in the future, but please keep it as it is :)  

cc @qsm-odoo

Comment (qsm-odoo) [reply addons/product/models/product_product.py]: @loco-odoo I am bit lost: if this is the current behavior, why do we need your diff here in `_set_template_field` to make it work? ^^" Or is it just about the "dirty" flag issue and then your diff would not be needed?

Comment (loco-odoo) [reply addons/product/models/product_product.py]: > Or is it just about the "dirty" flag issue and then your diff would not be needed?

Exactly, after the "dirty" flag issue fix, the diff in `_set_template_field` will not be needed anymore :+1:

Comment (paju-odoo) [reply addons/product/models/product_product.py]: Hi @loco-odoo,

Sorry for my late reply, I wasn't working yesterday... I'd prefer the latter option, have the template image displayed in the last position :)

**Comment by loco-odoo** [comment]
@qsm-odoo , this PR is now ready for your review :+1: Note that paju validated the behavior of it (among other things, the fact that the template image appears as the last image (even if it is removed) on the carousel for products with multiple variants and no specific image set on the main image but that have extra images).

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3793]: Mmh, indeed. Maybe we could do something more generic and just disable the observer during the call to _computeWidgetState (whatever the function is), no? And it should probably in 14.0/15.0 too?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1934]: Good catch :+1:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (loco-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: @qsm-odoo , here is the PR (https://github.com/odoo/odoo/pull/127606) for the fix in 15.0 (note that the problem is not there in 14.0 as the logic that handles the `o_dirty` class was a bit different). I removed the fix on this PR so the runbot will be red (as the test introduced in this PR does not benefit from all the fixes) but a soon as the other one is merged, I will rebase this one :+1:

**Comment by loco-odoo** [comment]
@qsm-odoo , just to be sure that you still have this one in mind :+1:

**Comment by loco-odoo** [comment]
@qsm-odoo , here is the rebased version.

**Review by qsm-odoo — APPROVED** [review]
Thanks for the fix :+1: 

@robodoo rebase-ff r+

**Pull request closed without merging**
